File: android/app/src/main/java/com/fyrbyadditive/famesmartblinds/data/local/AuthenticationManager.kt
package com.fyrbyadditive.famesmartblinds.data.local

import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Session expiry duration options
 */
enum class SessionExpiry(val displayName: String, val milliseconds: Long?) {
    HOUR("1 Hour", 3600_000L),
    DAY("1 Day", 86400_000L),
    WEEK("1 Week", 604800_000L),
    MONTH("1 Month", 2592000_000L),
    YEAR("1 Year", 31536000_000L),
    UNLIMITED("Unlimited", null)
}

/**
 * Represents an active authentication session for a device
 */
data class AuthSession(
    val deviceId: String,
    val authenticatedAt: Long,
    val expiresAt: Long?  // null means never expires
) {
    val isExpired: Boolean
        get() = expiresAt != null && System.currentTimeMillis() > expiresAt
}

/**
 * Manages authentication state for devices.
 * - Passwords are stored securely using EncryptedSharedPreferences
 * - Session metadata (expiry times) stored in regular SharedPreferences
 * - Provides a central point for checking/managing auth state
 */
@Singleton
class AuthenticationManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val secureStorage: SecureCredentialStorage
) {
    companion object {
        private const val TAG = "AuthenticationManager"
        private const val PREFS_NAME = "auth_sessions"
        private const val KEY_SESSIONS = "active_sessions"
        private const val KEY_DEFAULT_EXPIRY = "default_session_expiry"
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val gson = Gson()

    private val _activeSessions = MutableStateFlow<Map<String, AuthSession>>(emptyMap())
    val activeSessions: StateFlow<Map<String, AuthSession>> = _activeSessions.asStateFlow()

    private val _deviceNeedingAuth = MutableStateFlow<String?>(null)
    val deviceNeedingAuth: StateFlow<String?> = _deviceNeedingAuth.asStateFlow()

    var defaultSessionExpiry: SessionExpiry = SessionExpiry.WEEK
        private set

    init {
        loadSessions()
        loadDefaultExpiry()
        cleanExpiredSessions()
    }

    /**
     * Check if we have a valid session for a device
     */
    fun hasValidSession(deviceId: String): Boolean {
        val session = _activeSessions.value[deviceId] ?: return false
        return !session.isExpired
    }

    /**
     * Get the password for a device if we have a valid session
     */
    fun getPassword(deviceId: String): String? {
        if (!hasValidSession(deviceId)) return null
        return secureStorage.getPassword(deviceId)
    }

    /**
     * Authenticate with a device - stores password and creates session
     */
    fun authenticate(deviceId: String, password: String, expiry: SessionExpiry = defaultSessionExpiry) {
        // Save password securely
        secureStorage.savePassword(deviceId, password)

        // Create session
        val expiresAt = expiry.milliseconds?.let { System.currentTimeMillis() + it }

        val session = AuthSession(
            deviceId = deviceId,
            authenticatedAt = System.currentTimeMillis(),
            expiresAt = expiresAt
        )

        val sessions = _activeSessions.value.toMutableMap()
        sessions[deviceId] = session
        _activeSessions.value = sessions
        saveSessions()

        Log.d(TAG, "Authenticated device: $deviceId, expires: ${expiresAt?.let { Date(it) } ?: "never"}")
    }

    /**
     * Clear authentication for a device (e.g., on 401 or user logout)
     */
    fun clearAuthentication(deviceId: String) {
        secureStorage.deletePassword(deviceId)
        val sessions = _activeSessions.value.toMutableMap()
        sessions.remove(deviceId)
        _activeSessions.value = sessions
        saveSessions()
        Log.d(TAG, "Cleared authentication for device: $deviceId")
    }

    /**
     * Handle a 401 response - clears session and signals need for re-auth
     */
    fun handleAuthenticationRequired(deviceId: String) {
        clearAuthentication(deviceId)
        _deviceNeedingAuth.value = deviceId
    }

    /**
     * Clear the pending auth request
     */
    fun clearAuthenticationRequest() {
        _deviceNeedingAuth.value = null
    }

    /**
     * Update the default session expiry preference
     */
    fun setDefaultSessionExpiry(expiry: SessionExpiry) {
        defaultSessionExpiry = expiry
        prefs.edit().putString(KEY_DEFAULT_EXPIRY, expiry.name).apply()
    }

    // MARK: - Persistence

    private fun loadSessions() {
        val json = prefs.getString(KEY_SESSIONS, null) ?: return
        try {
            val type = object : TypeToken<Map<String, AuthSession>>() {}.type
            val sessions: Map<String, AuthSession> = gson.fromJson(json, type)
            _activeSessions.value = sessions
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load sessions", e)
        }
    }

    private fun saveSessions() {
        val json = gson.toJson(_activeSessions.value)
        prefs.edit().putString(KEY_SESSIONS, json).apply()
    }

    private fun loadDefaultExpiry() {
        val name = prefs.getString(KEY_DEFAULT_EXPIRY, null)
        if (name != null) {
            defaultSessionExpiry = try {
                SessionExpiry.valueOf(name)
            } catch (e: Exception) {
                SessionExpiry.WEEK
            }
        }
    }

    private fun cleanExpiredSessions() {
        val expiredDeviceIds = _activeSessions.value.filter { it.value.isExpired }.keys
        if (expiredDeviceIds.isEmpty()) return

        val sessions = _activeSessions.value.toMutableMap()
        for (deviceId in expiredDeviceIds) {
            secureStorage.deletePassword(deviceId)
            sessions.remove(deviceId)
        }
        _activeSessions.value = sessions
        saveSessions()
        Log.d(TAG, "Cleaned ${expiredDeviceIds.size} expired sessions")
    }
}

File: ios/FAME Smart Blinds/Services/AuthenticationManager.swift
import Foundation
import Combine

/// Session expiry duration options
enum SessionExpiry: String, CaseIterable, Identifiable {
    case hour = "1 Hour"
    case day = "1 Day"
    case week = "1 Week"
    case month = "1 Month"
    case year = "1 Year"
    case unlimited = "Unlimited"

    var id: String { rawValue }

    var timeInterval: TimeInterval? {
        switch self {
        case .hour: return 3600
        case .day: return 86400
        case .week: return 604800
        case .month: return 2592000  // 30 days
        case .year: return 31536000
        case .unlimited: return nil
        }
    }
}

/// Represents an active authentication session for a device
struct AuthSession: Codable {
    let deviceId: String
    let authenticatedAt: Date
    let expiresAt: Date?

    var isExpired: Bool {
        guard let expiresAt = expiresAt else { return false }
        return Date() > expiresAt
    }
}

/// Manages authentication state for devices.
/// - Passwords are stored securely in Keychain
/// - Session metadata (expiry times) stored in UserDefaults
/// - Provides a central point for checking/managing auth state
@MainActor
class AuthenticationManager: ObservableObject {
    static let shared = AuthenticationManager()

    private let keychain = KeychainManager.shared
    private let userDefaultsKey = "deviceAuthSessions"

    /// Currently active sessions (non-expired)
    @Published private(set) var activeSessions: [String: AuthSession] = [:]

    /// Default session expiry for new authentications
    @Published var defaultSessionExpiry: SessionExpiry = .week

    /// Device that needs authentication (set when 401 received)
    @Published var deviceNeedingAuth: String?

    /// Callback when authentication is needed
    var onAuthenticationNeeded: ((String) -> Void)?

    private init() {
        loadSessions()
        cleanExpiredSessions()
    }

    // MARK: - Session Management

    /// Check if we have a valid session for a device
    func hasValidSession(forDeviceId deviceId: String) -> Bool {
        guard let session = activeSessions[deviceId] else { return false }
        return !session.isExpired
    }

    /// Get the password for a device if we have a valid session
    func getPassword(forDeviceId deviceId: String) -> String? {
        guard hasValidSession(forDeviceId: deviceId) else { return nil }
        return keychain.getPassword(forDeviceId: deviceId)
    }

    /// Authenticate with a device - stores password and creates session
    func authenticate(deviceId: String, password: String, expiry: SessionExpiry? = nil) throws {
        let sessionExpiry = expiry ?? defaultSessionExpiry

        // Save password to keychain
        try keychain.savePassword(password, forDeviceId: deviceId)

        // Create session
        let expiresAt: Date?
        if let interval = sessionExpiry.timeInterval {
            expiresAt = Date().addingTimeInterval(interval)
        } else {
            expiresAt = nil
        }

        let session = AuthSession(
            deviceId: deviceId,
            authenticatedAt: Date(),
            expiresAt: expiresAt
        )

        activeSessions[deviceId] = session
        saveSessions()

        print("[Auth] Authenticated device: \(deviceId), expires: \(expiresAt?.description ?? "never")")
    }

    /// Clear authentication for a device (e.g., on 401 or user logout)
    func clearAuthentication(forDeviceId deviceId: String) {
        try? keychain.deletePassword(forDeviceId: deviceId)
        activeSessions.removeValue(forKey: deviceId)
        saveSessions()
        print("[Auth] Cleared authentication for device: \(deviceId)")
    }

    /// Handle a 401 response - clears session and signals need for re-auth
    func handleAuthenticationRequired(forDeviceId deviceId: String) {
        clearAuthentication(forDeviceId: deviceId)
        deviceNeedingAuth = deviceId
        onAuthenticationNeeded?(deviceId)
    }

    /// Clear the pending auth request
    func clearAuthenticationRequest() {
        deviceNeedingAuth = nil
    }

    // MARK: - Persistence

    private func loadSessions() {
        guard let data = UserDefaults.standard.data(forKey: userDefaultsKey),
              let sessions = try? JSONDecoder().decode([String: AuthSession].self, from: data) else {
            return
        }
        activeSessions = sessions
    }

    private func saveSessions() {
        guard let data = try? JSONEncoder().encode(activeSessions) else { return }
        UserDefaults.standard.set(data, forKey: userDefaultsKey)
    }

    private func cleanExpiredSessions() {
        let expiredDeviceIds = activeSessions.filter { $0.value.isExpired }.map { $0.key }
        for deviceId in expiredDeviceIds {
            try? keychain.deletePassword(forDeviceId: deviceId)
            activeSessions.removeValue(forKey: deviceId)
        }
        if !expiredDeviceIds.isEmpty {
            saveSessions()
            print("[Auth] Cleaned \(expiredDeviceIds.count) expired sessions")
        }
    }

    // MARK: - Session Expiry Settings

    /// Update the default session expiry preference
    func setDefaultSessionExpiry(_ expiry: SessionExpiry) {
        defaultSessionExpiry = expiry
        UserDefaults.standard.set(expiry.rawValue, forKey: "defaultSessionExpiry")
    }

    /// Load saved session expiry preference
    func loadDefaultSessionExpiry() {
        if let saved = UserDefaults.standard.string(forKey: "defaultSessionExpiry"),
           let expiry = SessionExpiry(rawValue: saved) {
            defaultSessionExpiry = expiry
        }
    }
}
